### PR TITLE
Add `Supermetrics` source, tasks, and flow

### PR DESCRIPTION
**Pull request by fdelgadodyvenia**
## Checklist

This PR:
- [ ] follows the guidelines laid out in `CONTRIBUTING.md`
- [ ] links relevant issue(s)
- [ ] adds/updates tests (if appropriate)
- [ ] adds/updates docstrings (if appropriate)
- [ ] adds an entry in `CHANGELOG.md`

Code Checker:
image

Test Coverage:
image

### CODE REVIEW

File: src/viadot/sources/supermetrics.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        -------
+        list of str
+            List of column names.
+
+        Raises
+        ------
+        ValueError
+            If the column names cannot be determined.
+        """
+        response: dict = self.to_json()
```
Comment (trymzet) [inline src/viadot/sources/supermetrics.py:215]: This seems quite suboptimal - you're calling the API twice, once just to get the columns, then another time to get the data itself. I think extracting the columns and the data should be done with 1 API call.

File: tests/unit/test_supermetrics.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                "paging": {
+                    "current_page": 1,  # Current page number if pagination is used
+                    "total_pages": 1,
+                    "total_results": 3
+                }
+            }
+
+    mock_method = mocker.patch('viadot.sources.supermetrics.Supermetrics.to_json')
+    mock_method.return_value = mock_response
+    mock_method = mocker.patch('viadot.sources.supermetrics.Supermetrics._get_col_names')
```
Comment (trymzet) [inline tests/unit/test_supermetrics.py:98]: See the comment above about 2 API calls; this is related - this shouldn't have to be mocked as I would assume such method doesn't need to perform an API call - it should accept a param such as `response: dict[str, Any]` and extract the column names from it.

File: src/viadot/orchestration/prefect/tasks/supermetrics.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    Parameters
+    ----------
+    query_params : dict
+        A dictionary containing the parameters for querying the Supermetrics API. These parameters
+        define what data to retrieve and how the query should be constructed.
+        
+    config_key : str, optional
+        The configuration key used to retrieve stored connection parameters and credentials for Supermetrics.
+        If provided, this key is used to fetch configuration settings from a configuration file or environment.
+        If not provided, the `credentials_secret` parameter must be used to supply credentials.
```
Comment (trymzet) [inline src/viadot/orchestration/prefect/tasks/supermetrics.py:36]: This doesn't even make sense. Just copy the description from other sources/tasks.

File: src/viadot/orchestration/prefect/tasks/supermetrics.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    MissingSourceCredentialsError
+        If neither `credentials_secret` nor `config_key` is provided, the function raises this error,
+        indicating that no valid credentials were supplied to access the Supermetrics API.
+
+    Notes
+    -----
+    - The function is decorated with Prefect's `@task` decorator, which allows it to be used within a Prefect flow.
+    - The task is configured to retry up to three times with a delay of 10 seconds between retries if it encounters
+      an error. The task will timeout if it runs longer than 60 minutes.
+    - This function assumes that the `Supermetrics` class is initialized with either credentials or a configuration key.
```
Comment (trymzet) [inline src/viadot/orchestration/prefect/tasks/supermetrics.py:58]: This is LLM vomit, remove pls.

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        A flag indicating whether to overwrite the file in ADLS if it already exists. 
+        If set to False and the file exists, an error will be raised.
+
+    supermetrics_credentials_secret : str, optional
+        The name of the secret in your secret management system that contains the Supermetrics API credentials. 
+        If not provided, the function will use credentials specified in the configuration.
+
+    supermetrics_config_key : str, optional
+        The configuration key for the Supermetrics API, used to retrieve stored query parameters and credentials. 
+        This key corresponds to a section in your configuration files.
```
Comment (trymzet) [inline src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py:62]: Better to copy an existing config key docstring as this is verbose and incorrect

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    adls_credentials_secret : str, optional
+        The name of the secret in your secret management system that contains the ADLS credentials. 
+        If not provided, the function will use credentials specified in the configuration.
+
+    adls_config_key : str, optional
+        The configuration key for ADLS, used to retrieve stored connection parameters and credentials.
+        This key corresponds to a section in your configuration files.
```
Comment (trymzet) [inline src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py:70]: Better to copy an existing config key docstring as this is verbose and incorrect

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    
+    Raises
+    ------
+    ValueError
+        If the `adls_path` is not provided and cannot be determined from the configuration.
+
+    Notes
+    -----
+    This function combines the data extraction and loading process into one step, 
+    simplifying the ETL workflow for Supermetrics data into ADLS.
```
Comment (trymzet) [inline src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py:90]: Bloat

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    kwargs : dict[str, Any], optional
+        Additional keyword arguments that will be passed to the `supermetrics_to_df` function. 
+        This allows further customization of the Supermetrics query.
+
+    Returns
+    -------
+    None
+        This function does not return any value. It performs an ETL operation where data is extracted 
+        from Supermetrics, transformed into a DataFrame, and loaded into Azure Data Lake Storage.
```
Comment (trymzet) [inline src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py:80]: Bloat

**Review by trymzet — CHANGES_REQUESTED** [review]
Added my comments

**Comment by trymzet** [comment]
@fdelgadodyvenia I assume this https://github.com/dyvenia/viadot/pull/671 can be closed?

File: src/viadot/sources/supermetrics.py (hunk 1 above)
Comment (fdelgadodyvenia) [reply src/viadot/sources/supermetrics.py:205]: Created an issue to solve this https://github.com/dyvenia/viadot/issues/1023

File: tests/unit/test_supermetrics.py (hunk 2 above)
Comment (fdelgadodyvenia) [reply tests/unit/test_supermetrics.py:98]: Created an issue to solve this https://github.com/dyvenia/viadot/issues/1023

File: src/viadot/orchestration/prefect/tasks/supermetrics.py (hunk 3 above)
Comment (fdelgadodyvenia) [reply src/viadot/orchestration/prefect/tasks/supermetrics.py]: done

File: src/viadot/orchestration/prefect/tasks/supermetrics.py (hunk 4 above)
Comment (fdelgadodyvenia) [reply src/viadot/orchestration/prefect/tasks/supermetrics.py]: done

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py (hunk 5 above)
Comment (fdelgadodyvenia) [reply src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py]: done

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py (hunk 6 above)
Comment (fdelgadodyvenia) [reply src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py]: done

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py (hunk 7 above)
Comment (fdelgadodyvenia) [reply src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py]: done

File: src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py (hunk 8 above)
Comment (fdelgadodyvenia) [reply src/viadot/orchestration/prefect/flows/supermetrics_to_adls.py]: done

File: tests/integration/test_supermetrics.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+from viadot.config import get_source_config
+from viadot.orchestration.prefect.flows import supermetrics_to_adls
+import pdb
+
+def test_flow_call():
+    google_ads_params = {
+        "ds_id": "AW",
+        "ds_user": "google@velux.com",
```
[the remark this answers is not in the document]
Comment (fdelgadodyvenia) [reply tests/integration/test_supermetrics.py]: done

File: tests/integration/test_supermetrics.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+            "Campaignname",
+            "Clicks",
+        ],
+        "max_rows": 1,
+    }
+
+    state = supermetrics_to_adls(
+        query_params = google_ads_params,
+        supermetrics_credentials_secret = "supermetrics",
+        adls_credentials_secret = "app-azure-cr-datalakegen2-dev",
```
[the remark this answers is not in the document]
Comment (fdelgadodyvenia) [reply tests/integration/test_supermetrics.py]: done

File: src/viadot/sources/supermetrics.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+2. Create a `Supermetrics` instance using the credentials and optional query
+   parameters.
+3. Use the `query` method to set query parameters and `to_df` to retrieve data
+   as a pandas DataFrame.
+
+API Documentation:
+------------------
+- API Documentation: https://supermetrics.com/docs/product-api-getting-started/
+- Usage Limits: https://supermetrics.com/docs/product-api-usage-limits/
+"""
```
[the remark this answers is not in the document]
Comment (fdelgadodyvenia) [reply src/viadot/sources/supermetrics.py]: done

File: src/viadot/sources/supermetrics.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+        if not self.query_params:
+            raise ValueError("Please build the query first")
+
+        params = {"json": json.dumps(self.query_params)}
+        headers = {"Authorization": f"Bearer {self.api_key}"}
+
+        response = handle_api_response(
+            url=self.API_ENDPOINT, params=params, headers=headers, timeout=timeout
+        )
+        return response.json()
```
[the remark this answers is not in the document]
Comment (fdelgadodyvenia) [reply src/viadot/sources/supermetrics.py:128]: Created an issue to solve this https://github.com/dyvenia/viadot/issues/1023

**Comment by trymzet** [comment]
Closing since there is a newer PR for this https://github.com/dyvenia/viadot/pull/1054

**Pull request closed without merging**